extensions/amp-a4a/0.1/signature-verifier.js

```diff
@@ -0,0 +1,207 @@
+/**
+ * Copyright 2017 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {signingServerURLs} from '../../../ads/_a4a-config';
+import {cryptoFor} from '../../../src/crypto';
+import {dev} from '../../../src/log';
+import {xhrFor} from '../../../src/xhr';
+
+/** @enum {number} */
+export const VerificationFailure = {
+  NO_FAULT: 1,
+  KEY_NOT_FOUND: 2,
+  SIGNATURE_MISMATCH: 3,
+};
+
+/**
+ * @param {!Window} win
+ * @return {!SignatureVerifier}
+ */
+export function signatureVerifierFor(win) {
+  const propertyName = 'AMP_FAST_FETCH_SIGNATURE_VERIFIER_';
+  if (!win[propertyName]) {
+    win[propertyName] = new SignatureVerifier(win);
+  }
+  return win[propertyName];
+}
+
+class SignatureVerifier {
```

Comment:
Working on adding a lot more documentation and tests to everything; it'll be up soon.